lib/buildConnectorsJson/generateSchemaFromJs.js

```diff
@@ -5,129 +5,106 @@
+	return schema;
 };
 
-// Separate function down here so it can recursively call itself
-function generateSchema(config, topLevel) {
+//Generates the whole schema for a given  (but not necessarily the schema of a object value)
+function generateSchema(rawSchema, topLevel) {
 
 	var schema = {
-    type: "object",
-    properties: {}
+	    type: "object",
+	    properties: {},
+		required: []
 	};
 
-	if (topLevel === true) {
-		schema.$schema = "http://json-schema.org/draft-04/schema#";
-	}
-
-
-	var inputKeys = _.keys(config);
+	//Advanced only need for top level
+	if (topLevel) schema.advanced = [];
 
-	// Set variables as required
-	schema.required = _.filter(inputKeys, function (key) {
-		var messageInput = config[key];
-		return (_.isObject(messageInput) && messageInput.required === true);
+	//Set required and advanced variables respectively
+	_.forEach(rawSchema, function (valObj, key) {
+		if (_.isObject(valObj)) {
+			if (valObj.required === true) schema.required.push(key);
+			if (topLevel && valObj.advanced === true) schema.advanced.push(key);
+		}
 	});
 
-	// Set variables as advanced
-	schema.advanced = _.filter(inputKeys, function (key) {
-		var messageInput = config[key];
-		return (_.isObject(messageInput) && messageInput.advanced === true);
+	// Add the variables in
+	_.each(rawSchema, function (rawObject, key) {
+		schema.properties[key] = generateSchemaObject(rawObject, key);
 	});
 
-	// Standard restriction: top level schema never allows for additional properties
-	schema.additionalProperties = false;
-
-	// Add the variables in
-	_.each(config, function (param, key) {
-		var newParam = {};
+	return schema;
 
-		// Recursively generate the schema for objects
-		if (param.type === 'object') {
-			newParam = generateSchema(param.properties || {});
-		}
+}
 
-		// Add in all the fluff
-		_.extend(newParam, _.pick(param, [
-			'type',
-			'name',
-			'title',
-			'description',
-			'default',
-			'enum',
-			'additionalProperties',
-			'additionalItems',
-			'format',
-			'timeout_millis',
-      'lookup'
-		]));
-
-		// Un camelcase the default jsonpath
-		if (param.defaultJsonPath) {
-			newParam.default_jsonpath = param.defaultJsonPath;
-		}
+//The generation of a schema object is seperate, so that selective recusion can be applied
+function generateSchemaObject(rawObject, title) {
 
-		// Make a pretty title if not explicitly defined
-		if (!newParam.title) {
-			newParam.title = prettyTitle(newParam.name || key);
-		}
+	var validObject = {};
 
-		// Default to no additional properties
-		if (newParam.type === 'object' && !_.isBoolean(newParam.additionalProperties) && !_.isObject(newParam.additionalProperties)) {
-			newParam.additionalProperties = false;
-		}
+	//If type is object, recurse via generateSchema and set additionalProperties
+	if (rawObject.type === 'object') {
+		validObject = generateSchema(rawObject.properties || {}, false); //Recurse from generateSchema
+		var aP = rawObject.additionalProperties;
```

Comment:
Hmm I get what you're doing here but this will make things more confusing for people coming back to review the code. I'd recommend making this `additionalProperties` instead - long descriptive variables are your friend because of autocomplete ❤️ .